File: 시스템/대시보드/웹/js/cases.js
// 스킬 검토 패널 — 스킬 설명/SKILL.md를 보고, 그 아래 케이스를 검토한다.
// §9.1: 자동 격리된 모순(conflict)을 전 스킬 통합 배너로 노출하고, 조건 좁혀 분기(branch)로 해소한다.
import { api } from "./api.js?v=20260702-13";

let selectedSkill = null;
let reviewIndex = {};   // skill name -> { conflicts, review }
let skillIndex = {};     // skill name -> list summary

function el(tag, cls, text) {
  const e = document.createElement(tag);
  if (cls) e.className = cls;
  if (text != null) e.textContent = text;
  return e;
}

export async function renderCases() {
  const listEl = document.getElementById("cases-skill-list");
  if (!listEl) return;
  await renderReviewBanner();   // 통합 검토큐 먼저(어떤 스킬에 모순이 있는지 인덱스 채움)
  await renderPromotionBacklog();  // 승격 대기 백로그(대표 원클릭 승격 — candidate 적체 해소)
  let skills = [];
  try {
    skills = await api.listSkills();
  } catch (e) {
    listEl.innerHTML = "";
    listEl.append(el("li", "empty", "스킬 목록 실패: " + e.message));
    return;
  }
  // 모순/검토가 있는 스킬을 맨 위로, 그다음 케이스 많은 순
  skills.sort((a, b) => {
    const ra = reviewIndex[a.name] || {}, rb = reviewIndex[b.name] || {};
    if ((rb.conflicts || 0) !== (ra.conflicts || 0)) return (rb.conflicts || 0) - (ra.conflicts || 0);
    if ((rb.review || 0) !== (ra.review || 0)) return (rb.review || 0) - (ra.review || 0);
    return (b.maturity?.cases || 0) - (a.maturity?.cases || 0);
  });
  listEl.innerHTML = "";
  skillIndex = {};
  if (!skills.length) {
    listEl.append(el("li", "empty", "스킬 없음"));
    return;
  }
  skills.forEach((s) => {
    skillIndex[s.name] = s;
    const li = el("li", "cases-skill-item");
    const m = s.maturity || {};
    const rv = reviewIndex[s.name] || {};
    const nameRow = el("div", "cases-skill-name", s.name);
    if (rv.conflicts) nameRow.append(el("span", "case-badge conflict", `모순 ${rv.conflicts}`));
    else if (rv.review) nameRow.append(el("span", "case-badge review", `검토 ${rv.review}`));
    li.append(nameRow);
    const meta = el("div", "cases-skill-meta");
    meta.textContent = m.is_new
      ? "신규 (케이스 0)"
      : `케이스 ${m.cases || 0}` + (m.warn_harmful ? ` · ⚠ harmful ${m.harmful}` : "");
    li.append(meta);
    if (s.name === selectedSkill) li.classList.add("active");
    li.onclick = () => {
      selectedSkill = s.name;
      renderCases();
    };
    listEl.append(li);
  });
  if (selectedSkill) renderSkillDetail(selectedSkill);
}

async function renderReviewBanner() {
  const banner = document.getElementById("cases-review-banner");
  reviewIndex = {};
  let data;
  try {
    data = await api.skillsReview();
  } catch (e) {
    if (banner) banner.hidden = true;
    return;
  }
  (data.skills || []).forEach((s) => { reviewIndex[s.skill] = { conflicts: s.conflicts, review: s.review }; });
  if (!banner) return;
  if (!data.total_review) {
    banner.hidden = true;
    banner.innerHTML = "";
    return;
  }
  banner.hidden = false;
  banner.innerHTML = "";
  const conf = data.total_conflicts || 0;
  const head = el("div", "review-banner-head",
    conf ? `⚠ 해소 대기 모순 ${conf}건 — 대표 판단 필요` : `검토 대기 ${data.total_review}건`);
  if (conf) head.classList.add("danger");
  banner.append(head);
  const chips = el("div", "review-banner-chips");
  (data.skills || []).forEach((s) => {
    const label = s.conflicts ? `${s.skill} · 모순 ${s.conflicts}` : `${s.skill} · 검토 ${s.review}`;
    const chip = el("button", "review-chip" + (s.conflicts ? " danger" : ""), label);
    chip.type = "button";
    chip.onclick = () => { selectedSkill = s.skill; renderCases(); };
    chips.append(chip);
  });
  banner.append(chips);
}

// ── 승격 대기 백로그 — candidate 적체(감사: 75/100)를 대표가 훑고 원클릭 승격/폐기 ──
async function renderPromotionBacklog() {
  const box = document.getElementById("cases-promotion-backlog");
  if (!box) return;
  let data;
  try {
    data = await api.skillsPromotions();
  } catch (e) {
    box.hidden = true;
    return;
  }
  const items = data.items || [];
  if (!items.length) {
    box.hidden = true;
    box.innerHTML = "";
    return;
  }
  box.hidden = false;
  box.innerHTML = "";
  const head = el("div", "promotion-backlog-head",
    `승격 대기 ${data.total}건${data.ready_count ? ` (준비완료 ${data.ready_count}건 ⭐)` : ""} — 승격하면 확정 규칙으로 모든 방에 적용됩니다`);
  box.append(head);
  const list = el("div", "promotion-backlog-list");
  // 준비완료·worked 순으로 이미 정렬돼 옴. 과부하 방지로 상위 30건만, 나머지는 개수 고지.
  items.slice(0, 30).forEach((it) => {
    const row = el("div", "promotion-backlog-row" + (it.ready_to_promote ? " ready" : ""));
    const mark = it.ready_to_promote ? "⭐ " : "";
    const pol = it.polarity === "failed" ? "⛔" : "✅";
    row.append(el("div", "promotion-backlog-body",
      `${mark}[${it.skill}] ${pol} ${it.condition} → ${it.instruction}` +
      (it.worked ? ` (worked ${it.worked})` : "")));
    const actions = el("div", "promotion-backlog-actions");
    const promoteBtn = el("button", "ghost small", "승격");
    promoteBtn.type = "button";
    promoteBtn.onclick = async () => {
      promoteBtn.disabled = true;
      try {
        await api.promoteCase(it.skill, it.case_id, "대표 승격(승격 대기 백로그)");
      } catch (e) { alert(`승격 실패: ${e.message || e}`); }
      renderCases();
    };
    const retireBtn = el("button", "danger small", "폐기");
    retireBtn.type = "button";
    retireBtn.onclick = async () => {
      retireBtn.disabled = true;
      try {
        await api.retireCase(it.skill, it.case_id, "대표 폐기(승격 대기 백로그)");
      } catch (e) { alert(`폐기 실패: ${e.message || e}`); }
      renderCases();
    };
    const openBtn = el("button", "ghost small", "열기");
    openBtn.type = "button";
    openBtn.onclick = () => { selectedSkill = it.skill; renderCases(); };
    actions.append(promoteBtn, retireBtn, openBtn);
    row.append(actions);
    list.append(row);
  });
  if (items.length > 30) {
    list.append(el("div", "promotion-backlog-more", `… 외 ${items.length - 30}건 (승격/폐기하면 이어서 표시)`));
  }
  box.append(list);
}

async function renderSkillDetail(skill) {
  const detail = document.getElementById("cases-detail");
  if (!detail) return;
  detail.innerHTML = "";
  detail.append(el("div", "cases-detail-title", `${skill} — 스킬`));
  const cached = skillIndex[skill] || {};
  let data, info, detailError = "";
  try {
    [info, data] = await Promise.all([api.skillDetail(skill), api.skillCases(skill)]);
  } catch (e) {
    try {
      data = await api.skillCases(skill);
    } catch (caseError) {
      detail.append(el("div", "empty", "스킬 로드 실패: " + caseError.message));
      return;
    }
    detailError = e.message;
  }
  if (selectedSkill !== skill) return;
  const meta = { ...cached, ...(info || {}) };
  detail.append(renderSkillSummary(meta, data, detailError));
  detail.append(el("div", "cases-help",
    "케이스는 이 스킬 아래에 쌓이는 경우의 수입니다. 만들어지면 다음부터 자동으로 사용되고, "
    + "신뢰가 쌓이면 자동 승격됩니다. 대표님은 잘못된 케이스만 삭제하시면 됩니다."));
  detail.append(el("div", "cases-section-title", "케이스"));
  if (detailError) {
    detail.append(el("div", "cases-load-note", "SKILL.md 로드 실패: " + detailError));
  }
  if (!data) {
    detail.append(el("div", "empty", "케이스 로드 실패"));
    return;
  }
  const cases = data.cases || [];
  const conv = {};
  (data.convergence || []).forEach((c) => (conv[c.case_id] = c));
  const byId = {};
  cases.forEach((c) => (byId[c.case_id] = c));
  if (!cases.length) {
    detail.append(el("div", "empty", "이 스킬엔 아직 케이스가 없습니다."));
    return;
  }

  const act = (label, cls, fn) => {
    const b = el("button", cls, label);
    b.type = "button";
    b.onclick = async () => {
      b.disabled = true;
      try {
        await fn();
        await renderCases();
      } catch (e) {
        alert(e.message);
        b.disabled = false;
      }
    };
    return b;
  };

  const condOf = (cid) => (byId[cid] ? (byId[cid].condition || cid) : cid);

  cases.forEach((c) => {
    const cv = conv[c.case_id] || {};
    const card = el("div", "case-card");
    card.dataset.status = c.status || "";

    const head = el("div", "case-card-head");
    head.append(el("span", "case-status", c.status || ""));
    head.append(el("span", "case-polarity " + (c.polarity || ""), c.polarity || ""));
    if (c.status === "conflict") head.append(el("span", "case-badge conflict", "모순 격리"));
    if (cv.ready_to_promote) head.append(el("span", "case-badge ready", "승격 준비"));
    if (cv.needs_review && c.status !== "conflict") head.append(el("span", "case-badge review", "검토 필요"));
    if (c.sensitivity === "confidential") head.append(el("span", "case-badge conf", "대외비"));
    card.append(head);

    card.append(el("div", "case-line", "상황: " + (c.condition || "")));
    card.append(el("div", "case-line strong", "지시: " + (c.instruction || "")));
    // 모순이면 무엇과 충돌하는지 보여준다(대표가 어느 쪽이 맞는지 판단할 근거)
    if (c.status === "conflict" && (c.conflicts_with || []).length) {
      const cw = el("div", "case-line conflict-with",
        "충돌 상대: " + c.conflicts_with.map(condOf).join(" / "));
      card.append(cw);
    }
    const confTxt = (cv.confidence == null) ? "-" : cv.confidence;
    // 사용 상태: 죽은 상태가 아니면 '다음에도 자동 사용됨'. 👍/👎는 에이전트 사용결과로 자동 집계.
    const inUse = !["superseded", "retired", "expired", "graduated"].includes(c.status || "");
    card.append(el("div", "case-metrics",
      `${inUse ? "✅ 자동 사용 중" : "⛔ 미사용"} · 👍 ${cv.worked || 0} · 👎 ${cv.harmful || 0} · 신뢰 ${confTxt}`
      + ` · ${c.status === "active" ? "승격됨" : "신뢰 쌓이면 자동 승격"} · 발의 ${c.proposed_by || "-"}`));

    const actions = el("div", "case-actions");
    if (c.status === "conflict") {
      // 격리된 모순: 조건 좁혀 분기(권장) / 이 쪽 폐기 / 그래도 이 쪽 채택
      actions.append(act("✂ 분기(조건 좁히기)", "primary small", () => doBranch(skill, c.case_id)));
      actions.append(act("🗑 삭제", "danger small", () => api.retireCase(skill, c.case_id, "대표 모순해소: 이 쪽 폐기")));
      actions.append(act("그래도 채택", "ghost small", () => api.promoteCase(skill, c.case_id, "대표 모순해소: 이 쪽 채택")));
    } else {
      // 주 액션: 잘못된 케이스 삭제(그만 쓰기). 나머지는 선택(보통 자동이라 손 댈 필요 없음).
      actions.append(act("🗑 삭제(그만 쓰기)", "danger small", () => api.retireCase(skill, c.case_id, "대표 삭제: 잘못된 케이스")));
      if (c.status !== "active") actions.append(act("지금 승격(선택)", "ghost small", () => api.promoteCase(skill, c.case_id, "대표 수동 승격")));
      if (c.status === "active" || c.status === "provisional_must") actions.append(act("강등(선택)", "ghost small", () => api.demoteCase(skill, c.case_id, "대표 강등")));
      actions.append(act("👍", "ghost small", () => api.caseEvent(skill, c.case_id, "worked")));
      actions.append(act("👎", "ghost small", () => api.caseEvent(skill, c.case_id, "harmful")));
    }
    card.append(actions);

    detail.append(card);
  });
}

function renderSkillSummary(info, data, detailError = "") {
  const box = el("div", "skill-summary");
  const head = el("div", "skill-summary-head");
  head.append(el("span", "skill-summary-name", info.name || data.skill || selectedSkill || ""));
  if (info.grade) head.append(el("span", "case-badge", info.grade));
  const maturity = data?.maturity || info.maturity || {};
  head.append(el("span", "case-badge", `케이스 ${maturity.cases || 0}`));
  if (maturity.version) head.append(el("span", "case-badge", `v${maturity.version}`));
  if (maturity.warn_harmful) head.append(el("span", "case-badge review", `harmful ${maturity.harmful || 0}`));
  box.append(head);

  const desc = el("div", "skill-description", info.description || "frontmatter description이 비어 있습니다.");
  box.append(desc);
  if (info.path) box.append(el("div", "skill-path", info.path));

  const source = el("details", "skill-source");
  const summary = el("summary", "skill-source-summary", "SKILL.md 보기");
  source.append(summary);
  const pre = el("pre", "skill-source-code");
  const code = el("code", "", detailError ? "SKILL.md를 읽을 수 없습니다." : (info.content || ""));
  pre.append(code);
  source.append(pre);
  box.append(source);
  return box;
}

// 모순을 '병합·삭제' 말고 조건 좁혀 분기(more-specific-wins). 대표가 좁힐 키워드를 직접 준다.
async function doBranch(skill, caseId) {
  const raw = window.prompt(
    "이 케이스를 어떤 상황에만 적용할지 좁혀 적으세요 (쉼표로 키워드, 예: VIP,긴급).\n" +
    "상대 케이스와 겹치지 않게 조건을 좁히면 둘 다 각자 상황에서 살아납니다.");
  if (raw == null) return;                       // 취소
  const keywords = raw.split(",").map((s) => s.trim()).filter(Boolean);
  if (!keywords.length) { alert("키워드를 최소 하나 입력하세요."); return; }
  await api.branchCase(skill, caseId, { keywords, note: raw.trim() },
    { rationale: "대표 분기: " + raw.trim() });
  await renderCases();
}

export function wireCases() {
  const refresh = document.getElementById("cases-refresh");
  if (refresh) refresh.onclick = () => renderCases();
  // 케이스 탭을 처음 누를 때 로드되도록
  document.querySelectorAll('.vtab[data-view="casesView"]').forEach((btn) => {
    btn.addEventListener("click", () => renderCases());
  });
}

File: 시스템/대시보드/서버/routers/cases.py
# -*- coding: utf-8 -*-
"""스킬 케이스 API — core.case_ledger 위의 얇은 HTTP 껍데기 (P-wire-C1').

대표가 대시보드에서 직접(진짜 대표 세션) candidate 케이스를 검토·승인·강등·retire 한다.
매니저 tick은 candidate만 발의하고, active/must 확정은 여기(실제 대표 인증)에서만 일어난다.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from core import case_ledger, skill_smith

router = APIRouter(prefix="/api/skills", tags=["skills"])


class CaseAction(BaseModel):
    by: str = "대표"
    rationale: str = ""
    method: str | None = None   # promote 전용: 기본 owner_approval(대표 직접 승인)


class CaseOutcome(BaseModel):
    event: str                  # worked | harmful | applied
    by: str = "대표"
    rationale: str = ""


class CaseBranch(BaseModel):
    """모순 격리(conflict)를 조건 좁혀 분기 해소. applies_when/does_not_apply_when 중 최소 1개 필요."""
    applies_when: dict | None = None
    does_not_apply_when: list | None = None
    restore_to: str | None = None     # 미지정 시 격리 전 상태(pre_conflict_status)
    by: str = "대표"
    rationale: str = ""


def _sdir(skill: str):
    sdir = case_ledger.skill_dir(skill)
    if not sdir:
        raise HTTPException(status_code=404, detail=f"스킬 없음: {skill}")
    return sdir


@router.get("")
def list_skills():
    """전 스킬 + 파생 성숙도(대표가 어떤 스킬에 케이스가 쌓였는지 본다)."""
    return skill_smith.list_skills()


@router.get("/review")
def review_all():
    """전 스킬 통합 검토큐 — 격리된 모순(conflict)·harmful·검토기한을 대표가 한눈에 본다.

    §9.1 자동 격리가 만든 conflict는 사람이 해소해야 비로소 안전 루프가 닫힌다.
    이 엔드포인트가 그 '보이게 함'을 담당한다(스킬별로 흩어진 review_queue를 한 곳에 모음).
    """
    skills_out = []
    total_conflicts = 0
    total_review = 0
    for s in skill_smith.list_skills():
        name = s.get("name", "")
        sdir = case_ledger.skill_dir(name)
        if not sdir:
            continue
        try:
            rq = case_ledger.review_queue(sdir)
        except case_ledger.CaseLedgerError:
            continue
        if not rq:
            continue
        conflicts = [x for x in rq if x.get("status") == "conflict"]
        total_conflicts += len(conflicts)
        total_review += len(rq)
        skills_out.append({
            "skill": name,
            "conflicts": len(conflicts),
            "review": len(rq),
            "items": rq,
        })
    skills_out.sort(key=lambda x: (x["conflicts"], x["review"]), reverse=True)
    return {"total_conflicts": total_conflicts, "total_review": total_review, "skills": skills_out}


@router.get("/{skill}")
def skill_detail(skill: str):
    """선택한 스킬의 frontmatter 설명과 SKILL.md 원문을 대시보드에 제공한다."""
    try:
        return skill_smith.skill_detail(skill)
    except skill_smith.SkillSmithError as exc:
        raise HTTPException(status_code=404, detail=str(exc))


@router.get("/{skill}/cases")
def list_cases(skill: str):
    sdir = _sdir(skill)
    cases = [c for c in case_ledger.read_cases(sdir) if c.get("status") not in case_ledger.DEAD_STATUSES]
    return {
        "skill": skill,
        "maturity": case_ledger.maturity(sdir),
        "review_queue": case_ledger.review_queue(sdir),
        "convergence": case_ledger.case_convergence(sdir),
        "evaluator_reliability": case_ledger.evaluator_reliability(sdir),
        "cases": cases,
    }


@router.post("/{skill}/cases/{case_id}/event")
def case_event(skill: str, case_id: str, body: CaseOutcome):
    """케이스 적용 결과 기록(대표가 '이거 잘됐다/안됐다' 마킹). worked/harmful → 수렴·검토큐에 반영.

    P1: 이벤트 기록은 신호일 뿐, 상태 전이(active 확정/강등)는 promote/demote(판단·승인)로만 일어난다.
    """
    _sdir(skill)
    try:
        case_ledger.record_case_event(skill, case_id, body.event, by=body.by, rationale=body.rationale)
        return {"ok": True}
    except case_ledger.CaseLedgerError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.post("/{skill}/cases/{case_id}/promote")
def promote(skill: str, case_id: str, body: CaseAction):
    _sdir(skill)
    try:
        return case_ledger.promote_case(
            skill, case_id, by=body.by,
            rationale=body.rationale or "대표 승인",
            method=body.method or "owner_approval",
        )
    except case_ledger.CaseLedgerError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.post("/{skill}/cases/{case_id}/demote")
def demote(skill: str, case_id: str, body: CaseAction):
    _sdir(skill)
    try:
        return case_ledger.demote_case(skill, case_id, by=body.by, rationale=body.rationale or "대표 강등")
    except case_ledger.CaseLedgerError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.post("/{skill}/cases/{case_id}/retire")
def retire(skill: str, case_id: str, body: CaseAction):
    _sdir(skill)
    try:
        return case_ledger.retire_case(skill, case_id, by=body.by, rationale=body.rationale or "대표 폐기")
    except case_ledger.CaseLedgerError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.post("/{skill}/cases/{case_id}/branch")
def branch(skill: str, case_id: str, body: CaseBranch):
    """conflict 케이스를 조건 좁혀 분기 해소(more-specific-wins). 병합·삭제 아님(설계 §3)."""
    _sdir(skill)
    resolution = {"case_id": case_id}
    if body.applies_when is not None:
        resolution["applies_when"] = body.applies_when
    if body.does_not_apply_when is not None:
        resolution["does_not_apply_when"] = body.does_not_apply_when
    if body.restore_to:
        resolution["restore_to"] = body.restore_to
    try:
        out = case_ledger.branch_conflict(skill, [resolution], by=body.by, rationale=body.rationale or "대표 분기 해소")
        return out[0] if out else {"ok": True}
    except case_ledger.CaseLedgerError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
